### PR TITLE
ISSUE-1088: Adds JsonF pattern functor

### PR DESCRIPTION
**Pull request by Baccata**
Adds a pattern-functor that reflects the Json ADT, as well as
fold/unfold methods.

This can be used in conjunction with recursion schemes libraries.

### CODE REVIEW

**Comment by Baccata** [comment]
Force-pushed for scalafix import order.

File: modules/core/shared/src/main/scala/io/circe/JsonF.scala [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * A pattern-functor reflecting the JSON datatype structure in a
+ * non-recursive way.
+ */
+sealed trait JsonF[+A]
+
+// format: off
+case object JNullF                                  extends JsonF[Nothing]
+case class JBooleanF(b: Boolean)                    extends JsonF[Nothing]
```
Comment (kaishh) [inline modules/core/shared/src/main/scala/io/circe/JsonF.scala]: final everything?

Comment (Baccata) [reply modules/core/shared/src/main/scala/io/circe/JsonF.scala]: Addressed in b6a8a4e

**Comment by travisbrown** [comment]
Thanks, @Baccata! I'm going to go ahead and merge this, and we can decide whether it belongs in core before the 0.12.0 release.

**Pull request merged**